src/Compilers/CSharp/Portable/Syntax/CSharpLineDirectiveMap.cs

```diff
@@ -21,69 +21,143 @@ public CSharpLineDirectiveMap(SyntaxTree syntaxTree)
+                                    mappedLine = unmappedLine;
+                                    mappedPathOpt = null;
+                                    state = PositionState.Unmapped;
+                                    break;
+
+                                case SyntaxKind.NumericLiteralToken:
+                                    // skip both the mapped line and the filename if the line number is not valid
+                                    if (!lineToken.ContainsDiagnostics)
+                                    {
+                                        tryGetOneBasedNumericLiteralValue(lineToken, ref mappedLine);
+                                        tryGetStringLiteralValue(directive.File, ref mappedPathOpt);
+                                        return new LineMappingEntry(unmappedLine, mappedLine, mappedPathOpt, PositionState.Remapped);
+                                    }
+                                    break;
+                                default:
+                                    throw ExceptionUtilities.UnexpectedValue(lineToken);
                             }
+                        }
 
-                            if (directive.File.Kind() == SyntaxKind.StringLiteralToken)
-                            {
-                                mappedPathOpt = (string?)directive.File.Value;
-                            }
+                        return new LineMappingEntry(unmappedLine, mappedLine, mappedPathOpt, state);
+                    }
 
-                            state = PositionState.Remapped;
+                case LineSpanDirectiveTriviaSyntax spanDirective:
+                    {
+                        LinePosition mappedStart = default;
+                        LinePosition mappedEnd = default;
+                        int? characterOffset = null;
+                        string? mappedPathOpt = null;
+
+                        if (tryGetPosition(spanDirective.Start, isEnd: false, ref mappedStart) &&
+                            tryGetPosition(spanDirective.End, isEnd: true, ref mappedEnd) &&
+                            tryGetOptionalCharacterOffset(spanDirective.CharacterOffset, ref characterOffset) &&
+                            tryGetStringLiteralValue(spanDirective.File, ref mappedPathOpt))
+                        {
+                            return new LineMappingEntry(unmappedLine, new LinePositionSpan(mappedStart, mappedEnd), characterOffset, mappedPathOpt);
                         }
 
-                        break;
+                        return new LineMappingEntry(unmappedLine, unmappedLine, mappedPathOpt: null, PositionState.Unmapped);
+                    }
+
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(directive);
+            }
+
+            static bool tryGetOneBasedNumericLiteralValue(in SyntaxToken token, ref int value)
+            {
+                if (!token.IsMissing &&
```

Comment:
It looks like these two checks (`IsMissing` and `Kind() == NumericLiteralToken`) were not in the original source code. Do we need them?

Reply:
> It looks like these two checks

The original checks are still there above the call to the local function, so these are redundant.